src/multihash.rs

```diff
@@ -173,6 +173,45 @@ impl<const S: usize> Multihash<S> {
             .expect("writing to a vec should never fail");
         bytes
     }
+
+    /// Truncates the multihash to the given size. It's up to the caller to ensure that the new size
+    /// is safe to use.
+    ///
+    /// If the new size is larger than the current size, this method does nothing.
+    ///
+    /// ```
+    /// use multihash::{Code, MultihashDigest};
+    ///
+    /// let hash = Code::Sha3_256.digest(b"Hello world!").truncate(20);
+    /// ```
+    pub fn truncate(&self, size: u8) -> Self {
```

Comment:
since its up to the caller to ensure the size is safe to use, this should be marked `unsafe`. while not quite UB as somewhere it would probably error, Sha2_512 truncated into 32 bytes is something that isnt exactly great. A side note for this, we arent doing any checks to ensure the size of what we read matches with the expected size of the hash, which I will open probably an issue for.

Reply:
actually the issue I stated above is currently avoided as we dont have any way of converting a multihash into the digest type. Though reading and creation of malformed multihashes from external sources is allowed.

Reply:
"unsafe" is for rust-level unsafety. The new digest may not be secure, but not all of our hash functions are "secure" anyways. That kind of security needs to be enforced through a list of allowed hash functions/sizes.

Reply:
I was not talking about the security of the hash functions. What I am talking about is the allowance of Multihash structs to be formed from non-spec conforming data, every function I saw that converts bytes into a multihash performs no checks outside of if it fits within buffer and struct bounds. 
The unsafe keyword is a good way to inform the user that using this can easily cause malformed data and/or UB if not properly handled, although this wont cause any memory errors or UB since its all safe rust, easily produces incorrect data if used incorrectly and because of that I think it should be labeled `unsafe`.
The `size` field is used to read and write multihashes stuff like the [example here](https://github.com/multiformats/rust-multihash/issues/157) is considered a valid multihash going in and out which is not good.

IMO at the very least a better wording of this function would be `truncate_unchecked` to at least express the need for the user to check what they are doing or at least read the documentation.

Reply:
really this should all be checked when decoding to prevent malformed data which is why I created an issue, but since its not checked when encoding either, this function can/will allow for malformed Multihashes to be produced by this lib.

And again, there is no risk of UB or segfaults here, just the risk of outputting multihashes that dont make sense.

Reply:
I like the idea of having it called `truncate_unchecked`. I prefer if libraries try hard to produce valid data.

Reply:
This isn't unchecked and it always produces _valid_ multihashes. It just might not produce _secure_ hash digests.

> The unsafe keyword is a good way to inform the user that using this can easily cause malformed data and/or UB if not properly handled, although this wont cause any memory errors or UB since its all safe rust, easily produces incorrect data if used incorrectly and because of that I think it should be labeled unsafe.

Please don't use unsafe for this. That's not what it means and it makes code difficult to read/audit.

Reply:
I have updated the documentation to clarify the safety (safe -> secure).

Reply:
> What I am talking about is the allowance of Multihash structs to be formed from non-spec conforming data

To be clear, the multihash spec allows multihashes to be arbitrarily truncated. I.e., a length-20 SHA256 multihash is a SHA256 multihash truncated to 20 bytes.